kizhin.evgeniy/T1

sdjklfjsdlkfjsdklfj opened the pull request:
REGULAR | POLYGON
12 баллов

Commit:
Initial commit for T1: implement essential clases

Commit:
Implement basic utils

Implemented: input parsing, shape creation, scaling logic,
output formating

Commit:
Implement main

main.cpp with main function acting according the task and
main_utils modules that help to avoid code repating and simplify
main function in favour of readability

Commit:
Implement Polygon Shape

Commit:
Implemet Regular Shape and point_utils

Commit:
Refactor: Revise the entire code base

Extensive refactoring of the code base to improve code cleanliness
and efficiency. Revised functions and class fields, optimized data
manipulation.

Commit:
Minor refactoring

Commit:
Refactor colons and class initialization lists

Volgarenok commented:
@sdjklfjsdlkfjsdklfj индивидуальное задание: реализуйте `REGULAR` через `POLYGON`

Commit:
Refactor: Replace C-style code with C++

Replaced C-style macroses with constepr
Replaced C-style strings with std::string
Replaced return codes with exceptions
Minor refactoring

Commit:
Refactor: remove unused functions, simplify shapes

- Moved scaling factor checks to shape classes from shape_utils
- Removed extra asserts, used exceptions in some places
- Removed size passings to functions - used nullptr as shapes array
  invariant
- Minor changes

Commit:
Refactor: CG 5, 48

Commit:
Update IO Utilities

Fixed issue with extra stream checking that leads to unexpected
error

Commit:
Refactor: spaces in Braced Lists

Commit:
Implement Regular with Polygon

Commit:
Introduce CompositeShape and virtual copy constructors

Volgarenok commented on kizhin.evgeniy/T1/io_utils.cpp:
```diff
@@ -0,0 +1,112 @@
+  try {
+    while (is) {
+      currentParams = parseInputShape(is, currentName);
+      if (currentName == "SCALE") {
+        isScaleCommand = true;
+        break;
+      }
```
Это тоже условие выхода из цикла -> должно быть в while

Volgarenok commented on kizhin.evgeniy/T1/main.cpp:
```diff
@@ -0,0 +1,24 @@
+int main()
+{
+  constexpr size_t shapesCapacity = 10'000;
+  kizhin::Shape* shapes[shapesCapacity]{};
```
Здесь локально в main можно использовать конструкцию `using namespace <ваш namespace>`, потому что часто к нему обращаемся

Volgarenok commented on kizhin.evgeniy/T1/rectangle.cpp:
```diff
@@ -0,0 +1,38 @@
+
+void kizhin::Rectangle::scale(double scaleFactor)
+{
+  if (scaleFactor <= 0.0) {
+    throw std::logic_error("Failed to scale");
+  }
```
Проверка дублируется в классах-наследниках. Следует использовать паттерн "Шаблонный метод" реализованный через идиому невиртуального интерфейса (см. Мейерс "55 советов...", Глава 6)

Volgarenok commented on kizhin.evgeniy/T1/shape_utils.cpp:
```diff
@@ -0,0 +1,104 @@
+
+void kizhin::scaleShapes(Shape* const* shapes, const double* params)
+{
+  for (Shape* const* i = shapes; *i != nullptr; ++i) {
+    scaleShape(*i, params);
+  }
```
Вот тут-то вы уже всё знаете про параметры. Во всяком случае вы можете их проверить. Здесь как раз не хватает второго метода. Условного `unsafeScale`, который пользователи могут использовать, если "в себе уверены"

Volgarenok reviewed (commented):
Норм

Commit:
Merge master of Volgarenok/spbspu-labs-2024-aap-a

Commit:
Merge master of Volgarenok/spbspu-labs-2024-aap-a

Commit:
Merge kizhin.evgeniy/T1-part2 into kizhin.evgeniy/T1

Commit:
Refactor Shapes: NVI for scale methods

- Removed duplicate checks in scale methods with NVI
- Refactored loop in io_utils
- Added using namespace to main (idk why)

Commit:
Update exception messages

Commit:
Add unsafe scaleShapes vesion. Refactor exceptions

Commit:
Implement clone in CompositeShape

Commit:
Fix shape leak in IO utils

Volgarenok commented on kizhin.evgeniy/T1/composite_shape.cpp:
```diff
@@ -0,0 +1,236 @@
+}
+
+kizhin::CompositeShape::CompositeShape(size_t capacity):
+  begin_(nullptr),
+  end_(nullptr),
+  end_cap_(nullptr)
+{
+  if (capacity > 0) {
+    begin_ = new Shape*[capacity];
+    end_ = begin_;
+    end_cap_ = begin_ + capacity;
```
Тернарный оператор в помощь

Volgarenok commented on kizhin.evgeniy/T1/shape_utils.cpp:
```diff
@@ -0,0 +1,101 @@
+  return new Regular{ p1, p2, p3 };
+}
+
+kizhin::Polygon* kizhin::createPolygon(const double* params)
+{
+  const size_t size = static_cast< size_t >(params[0]);
+  const size_t pointCount = size / 2;
+  point_t* points = new point_t[pointCount];
+  for (size_t i = 0; i < pointCount; ++i) {
+    points[i].x = params[1 + i * 2];
+    points[i].y = params[2 + i * 2];
+  }
+  Polygon* result = nullptr;
+  try {
+    result = new Polygon(points, pointCount);
```
Неудачно, потому что вы уже в трёх местах сохранили точки: при вводе, здесь, да ещё и в полигоне. Надо что-то другое придумать

Volgarenok commented on kizhin.evgeniy/T1/shape_utils.cpp:
```diff
@@ -0,0 +1,101 @@
+{
+  const point_t leftDown{ params[1], params[2] };
+  const point_t rightUp{ params[3], params[4] };
+  if (leftDown.x >= rightUp.x || leftDown.y >= rightUp.y) {
+    throw std::logic_error("Invalid pointns to create rectangle");
+  }
```
Наверное это ответственность конструктора `Rectangle`

Volgarenok replied:
Ну во всяком случае у вас на нём никакой ответственности. Может и конструктор другой нужен?

Volgarenok commented on kizhin.evgeniy/T1/shape_utils.cpp:
```diff
@@ -0,0 +1,101 @@
+  Polygon* createPolygon(const double*);
+}
+
+void kizhin::scaleShape(Shape* shape, double scalingFactor, const point_t& scalingPoint)
+{
+  const point_t oldFramePos = shape->getFrameRect().pos;
+  shape->move(scalingPoint);
+  const point_t newFramePos = shape->getFrameRect().pos;
+  shape->scale(scalingFactor);
```
А какая тут гарантия безопасности? Может получше можно?

Volgarenok replied:
Слишком радикально, ну да ладно

Volgarenok commented on kizhin.evgeniy/T1/shape.hpp:
```diff
@@ -0,0 +1,27 @@
+
+  private:
+    virtual void scaleWithoutChecks(double) = 0;
+    void throwIfInvalidScalingFractor(double);
```
Наверное это свободная функция в файле с реализацией всё-таки, а не метод

Volgarenok commented on kizhin.evgeniy/T1/regular.cpp:
```diff
@@ -0,0 +1,85 @@
+#include <stdexcept>
+#include "point_utils.hpp"
+
+kizhin::Regular::Regular(const point_t& p1, const point_t& p2, const point_t& p3)
+{
+  if (!isRightTriangle(p1, p2, p3)) {
+    throw std::invalid_argument("Invalid triangle for regular construction");
+  }
+  const double r1 = computeDistance(p1, p2);
+  const double r2 = computeDistance(p1, p3);
+  const double outerRadius = std::max(r1, r2);
+  const double innerRadius = std::min(r1, r2);
+  const point_t vertex = (outerRadius == r1) ? p2 : p3;
+  const size_t size = computeSize(innerRadius, outerRadius);
+  const point_t* vertices = nullptr;
+  try {
+    vertices = computeVerticesArray(p1, vertex, size);
+    polygon_ = Polygon{ vertices, size };
+  } catch (...) {
+    delete[] vertices;
+    throw;
+  }
+  delete[] vertices;
+}
+
```
Надо проинициализировать в списке инициализации. Для этого надо сделать функцию и вызвать её там

Volgarenok commented on kizhin.evgeniy/T1/shape.hpp:
```diff
@@ -0,0 +1,27 @@
+
+    virtual void move(const point_t&) = 0;
+    virtual void move(double, double) = 0;
+    void scale(double);
+
+  private:
+    virtual void scaleWithoutChecks(double) = 0;
```
Ну а вдруг я знаю где-то заранее, что коэффициент положительный и проверять я его не хочу?

Volgarenok commented on kizhin.evgeniy/T1/polygon.hpp:
```diff
@@ -0,0 +1,41 @@
+  class Polygon final: public Shape
+  {
+  public:
+    Polygon(): vertices_(nullptr), size_(0) {}
```
Ну полигон по умолчанию - это прям плохо. Не должно его быть в принципе

Volgarenok commented on kizhin.evgeniy/T1/polygon.cpp:
```diff
@@ -0,0 +1,169 @@
+  return false;
+}
+
+size_t
+kizhin::countEqual(const point_t* begin, const point_t* end, const point_t val)
```
Бомбический перенос, конечно

Volgarenok commented on kizhin.evgeniy/T1/point_utils.cpp:
```diff
@@ -0,0 +1,65 @@
+  return std::sqrt(dx * dx + dy * dy);
+}
+
+double* kizhin::computeEdgeCords(const point_t* points, size_t size)
```
Конкретно здесь разрешаю использовать `std::array` вместо динамического массива

Volgarenok commented on kizhin.evgeniy/T1/polygon.cpp:
```diff
@@ -0,0 +1,169 @@
+
+kizhin::Polygon& kizhin::Polygon::operator=(Polygon&& rhs) noexcept
+{
+  if (this != &rhs) {
```
Для взятия адреса используйте `std::addressof`

Volgarenok commented on kizhin.evgeniy/T1/composite_shape.hpp:
```diff
@@ -0,0 +1,49 @@
+    Shape* operator[](size_t) noexcept;
+    const Shape* operator[](size_t) const noexcept;
+
+    size_t size() const noexcept { return end_ - begin_; }
+    bool empty() const noexcept { return begin_ == end_; }
```
Да не нужно это. Оставьте в `cpp`. Компановщик умеет в инлайнинг, если его очень сильно попросить

Volgarenok commented on kizhin.evgeniy/T1/composite_shape.cpp:
```diff
@@ -0,0 +1,236 @@
+  --end_;
+}
+
+void kizhin::CompositeShape::resize(size_t newSize)
+{
+  if (newSize <= size()) {
+    for (Shape** i = begin_ + newSize; i != end_; ++i) {
+      delete *i;
+    }
+    end_ = begin_ + newSize;
+    return;
+  }
+  const size_t currentSize = size();
+  Shape** newBegin = new Shape*[newSize];
+  copy(begin_, begin_ + currentSize, newBegin);
+  clear();
+  begin_ = newBegin;
+  end_ = begin_ + currentSize;
+  end_cap_ = begin_ + newSize;
+}
```
Мне кажется можно получше сделать с  "copy-and-swap"

Volgarenok commented on kizhin.evgeniy/T1/composite_shape.cpp:
```diff
@@ -0,0 +1,236 @@
+
+kizhin::CompositeShape* kizhin::CompositeShape::clone() const
+{
+  CompositeShape* newComposite = new CompositeShape();
+  for (const Shape* const* i = begin_; i != end_; ++i) {
+    newComposite->push_back((*i)->clone());
+  }
+  return newComposite;
+}
```
Исключение в `push_back` - утечка склонированной фигуры (:

Аккуратней ннннада (:

Volgarenok replied:
Ну и в целом `newComposite` утекает при любом исключении

Volgarenok commented on kizhin.evgeniy/T1/composite_shape.cpp:
```diff
@@ -0,0 +1,236 @@
+}
+
+kizhin::CompositeShape&
+kizhin::CompositeShape::operator=(CompositeShape&& rhs) noexcept
+{
+  if (this != &rhs) {
+    clear();
```
В полигоне бы как-то так же сделали, но `clear` закрытым

Commit:
Introduce CopyAssign, Update Rectangle constructor

Refactored a lot of code as well

Commit:
Fixed leak in composite_shape clone method

Commit:
Used initialization in regular constructor

Commit:
Reduce allocations count during Polycon construction

Commit:
Refactor exceptions

Commit:
Add return type to scaleShape. Remove copyAssign

Commit:
Refactor duplicates search logic in Polygon

Commit:
Merge master of Volgarenok/spbspu-labs-2024-aap-a into kizhin.evgeniy/T1

Commit:
Update clone: use copy constructor

Commit:
Merge master of Volgarenok/spbspu-labs-2024-aap-a into kizhin.evgeniy/T1

Commit:
Merge master of Volgarenok/spbspu-labs-2024-aap-a into kizhin.evgeniy/T1

Commit:
Remove copy from composite_shape implementation

Copy function leads to memory leaks when some on clone class throw
exception, so end_ is updated correctly

Commit:
Update pop_back: fix memory leak

Commit:
Refactor scale: strong exception safety via copy&swap

Commit:
Add params count checks to makeSmth functions

Commit:
Refactor move: strong exception safety via copy&swap
